diff --git '${BUILD_DIR}'uboot-linux4sam_5.7/include/configs/sama5d2_xplained.h '${BUILD_DIR}'uboot-linux4sam_5.7/include/configs/sama5d2_xplained.h
index ed30e9b..c7f17e5 100644
--- '${BUILD_DIR}'uboot-linux4sam_5.7/include/configs/sama5d2_xplained.h
+++ '${BUILD_DIR}'uboot-linux4sam_5.7/include/configs/sama5d2_xplained.h
@@ -45,6 +45,8 @@
 /* LCD */
 #ifdef CONFIG_ATMEL_HLCD
 #define CONFIG_SYS_WHITE_ON_BLACK
+#define CONFIG_CMD_BMP
+#define CONFIG_SYS_MEM_TOP_HIDE 	(1 << 20)
 #endif
 
 #ifdef CONFIG_SYS_USE_MMC
@@ -53,13 +55,13 @@
 #undef FAT_ENV_DEVICE_AND_PART
 #undef CONFIG_BOOTCOMMAND
 
-#define FAT_ENV_DEVICE_AND_PART	"1"
-#define CONFIG_BOOTCOMMAND	"fatload mmc 1:1 0x21000000 at91-sama5d2_xplained.dtb; " \
-				"fatload mmc 1:1 0x22000000 zImage; " \
+#define FAT_ENV_DEVICE_AND_PART	"0"
+#define CONFIG_BOOTCOMMAND	"fatload mmc 0:1 0x22000000 sps.bmp; bmp display 0x22000000; fatload mmc 0:1 0x21000000 evdc1b_emmc_wu.dtb; " \
+				"fatload mmc 0:1 0x22000000 zImage; " \
 				"bootz 0x22000000 - 0x21000000"
 #undef CONFIG_BOOTARGS
 #define CONFIG_BOOTARGS \
-	"console=ttyS0,115200 earlyprintk root=/dev/mmcblk1p2 rw rootwait"
+      "console=ttyS0,115200 root=/dev/mmcblk0p2 rw rootfstype=ext4 rootwait video=Unknown-1:800x480-16 atmel.pm_modes=standby,ulp1"
 
 #endif
